celavi/pylca_celavi/pylca_opt_background.py

```diff
@@ -423,55 +121,68 @@ def runner(tech_matrix, F,i,l,j,k,route_id,final_demand_scaler):
-           pass
+          print(f"pylca-opt-background emission failed for {k} at {j} in {i}", flush = True) 
+          pass       
+        
     
         print(str(time.time() - tim0) + ' ' + 'taken to do this run',flush=True)
     
         return res
     
 
-    
+    process_emissions = pd.read_csv(uslci_emission_filename)
+    #Creating the technology matrix for performing LCA caluclations
+    tech_matrix = pd.read_csv(uslci_tech_filename,index_col ='conjoined_flownames')
+    process_adder = pd.read_csv(uslci_process_filename)
+    process_adder['product'] = process_adder['product'].str.lower()
+    f_d['flow name'] = f_d['flow name'].str.lower()
+    f_d = f_d.merge(process_adder, left_on = "flow name", right_on = "product")
+    #This list of products and processes essentially help to determine the indexes and the products and processes
+    #to which they belong. 
+    uslci_products = pd.DataFrame(tech_matrix.index)
+    uslci_process = list(tech_matrix.columns)
+
+
     #Have to edit this final demand to match the results from CELAVI
-    f_d['flow name'] = f_d['flow name'] +'@' + f_d['flow name']
+    #f_d['flow name'] = f_d['flow name'] +'@' + f_d['flow name']
+    #f_d['flow name'] = f_d['flow name'].str.lower()
     f_d = f_d.drop_duplicates()
     f_d = f_d.sort_values(['year'])
 
 
     #Replace electricity    
-    f_d  = f_d .replace(to_replace='electricity@electricity', value='electricity, at grid@electricity, at grid, us, 2010')
+    #f_d  = f_d .replace(to_replace='electricity@electricity', value='electricity, at grid, us, 2010@electricity, at grid, us, 2010')
+    #f_d['flow name'] = f_d['flow name'] +'@' + f_d['location']
+    #f_d['flow name'] = f_d['flow name'].str.lower()
+    f_d['conjoined_flownames'] = f_d['conjoined_flownames'].str.lower()  
+    f_d.to_csv('../../celavi-data/generated/intermediate_demand_foreground_uslci_data.csv', mode = 'a')
+    #dataframe to debug connecting between foreground and background  
+    uslci_products['conjoined_flownames'] = uslci_products['conjoined_flownames'].str.lower()  
+    final_dem = uslci_products.merge(f_d, left_on='conjoined_flownames', right_on='conjoined_flownames', how='left')
+    final_dem = final_dem.fillna(0)
 
-    uslci_product_df[0] = uslci_product_df[0].str.lower()
-    f_d['flow name'] = f_d['flow name'].str.lower()
 
-    final_dem = uslci_product_df.merge(f_d, left_on=0, right_on='flow name', how='left')
-    final_dem = final_dem.fillna(0)
     chksum = np.sum(final_dem['flow quantity'])
-    #To make the optimization easier
+    if chksum == 0:
+        print('Final demand construction failed. No value. csv file for error checking created. /n Check the intermediate demand file and final demand check file')
+        final_dem.to_csv('../../celavi-data/generated/Final_demand_check_file.csv')
+        uslci_products.to_csv('../../celavi-data/generated/uslciproducts_file.csv')
+        f_d.to_csv('../../celavi-data/generated/demandforeground.csv')
```

Comment:
Lines 180-182: Remove all hard-coded file directories and names, replace with parameters in the YAML file(s)